### PR TITLE
Additional updates for split grid simulations

### PR DESCRIPTION
**Pull request by alfrob98**
Have added new BCs, new subgrid routines, and new subgrid IO routines. All unit tests up to date.

### CODE REVIEW

**Comment by alfrob98** [comment]
In the most recent commit, I've added a correction to the 1D vR calculation that assumes a gaussian form of the dust vertical density distribution and integrates over the 2D gas density and velocity profile to correctly calculate the dust radial velocity when it is coupled to the gas. With the correction, there is now good agreement of the integrated radial velocities at the subgrid boundary (see fig.). 

vRdust


Given one needs the 2D gas distribution, this correction is only currently usable if one is running a split grid simulation where the gas density and velocities are calculated over the whole grid (which is my standard split grid set up at the moment).

**Comment by rbooth200** [comment]
I noticed that you added more opacity files to the repository. This is ok, but I wonder if it's better to provide scripts that generate them to avoid having a large number of big files. I wouldn't worry about removing these ones.

File: src/DSHARP_opacs.cu
Comment (rbooth200) [inline src/DSHARP_opacs.cu:71]: Is there a reason this function does not also compute rho_tot? I thought rho_tot was needed for some of the radiation routines. This seems like a natural place to do it.

**Review by rbooth200 — COMMENTED** [review]
This is looking good. Just a couple of questions. Also, at some point in time the code could probably use some more documentation - especially for these complex parts of the code.

**Comment by alfrob98** [comment]
> I noticed that you added more opacity files to the repository. This is ok, but I wonder if it's better to provide scripts that generate them to avoid having a large number of big files. I wouldn't worry about removing these ones.

Good idea. Will neaten up the script I have for that and add in at some point.

File: src/DSHARP_opacs.cu
Comment (alfrob98) [reply src/DSHARP_opacs.cu:71]: Yeah good point, will make these changes now.

**Comment by alfrob98** [comment]
I've added the rho_tot calculation to the rhokappa tot calculation for the functions where the user inputs a prims object for the dust density. I haven't added it to the case where the density is passed as a field of doubles as that case is specific to the split grid routine where (currently) the user needs to define how they want the density to be mixed between 1D and 2D grids. 

I haven't pushed the split grid cpp file that I've been working on for all of the testing, but can add it if you think its useful? Or if you just want to see it I can send you the file.

**Comment by rbooth200** [comment]
My feeling about the split grid stuff is that it is pretty complex, and without a working example, it won't be possible for someone else to get it working. However, I'd only upload an example now if you're happy these routines are done and won't need adapting in the future. Otherwise, it's better to keep testing/using it and to upload what we think is the best way to do it when we're sure. At that stage, we'd want to provide the routines to mix the grids along with a working example. It would be good to tidy up the split grid code and remove any duplication that's crept in at that time rather than worrying about it here.

**Comment by alfrob98** [comment]
Yeah I agree with all that. Do you therefore think it best not to merge it as it is now, and wait until later? Or are you happy to merge now with it in its current form, it being a feature that only we use?

**Comment by alfrob98** [comment]
The reason I ask is that I want to start setting up the ice-vapour solver on the split grid, but I guess I can merge the commits from the splitgridsim branch into the icevap branch instead of merging from main.

**Comment by rbooth200** [comment]
I will merge this as-is then.

**Review by rbooth200 — APPROVED** [review]

**Pull request merged**
